cpp/include/cuspatial/utility/device_atomics.cuh

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace cuspatial {
+namespace detail {
+
+__device__ double atomicMin(double* addr, double val)
+{
+  unsigned long long int* address_as_ull = (unsigned long long int*)addr;
+  unsigned long long int old             = *address_as_ull, assumed;
+
+  do {
+    assumed = old;
+    old     = atomicCAS(
+      address_as_ull, assumed, __double_as_longlong(std::min(val, __longlong_as_double(assumed))));
+
+    // Note: uses integer comparison to avoid hang in case of NaN (since NaN != NaN)
+  } while (assumed != old);
+
+  return __longlong_as_double(old);
+}
```

Comment:
1. Access `addr` as `double` before conversion to avoid strict aliasing violation (the other aliased accesses are using compiler intrinsics so any violations are harmless).
2. Remove c-style cast
3. Don't declare two variables on a single line.
4. Repeat for the other implementation below.

(Thanks @jrhemstad)

```
__device__ double atomicMin(double* addr, double val)
{
  double old_double = *addr;
  unsigned long long int* address_as_ull = reinterpret_cast<unsigned long long int*>(addr);
  unsigned long long int old   = __double_as_longlong(old_double);
  unsigned long long int assumed;

  do {
    assumed = old;
    old     = atomicCAS(
      address_as_ull, assumed, __double_as_longlong(std::min(val, __longlong_as_double(assumed))));

    // Note: uses integer comparison to avoid hang in case of NaN (since NaN != NaN)
  } while (assumed != old);

  return __longlong_as_double(old);
}
```